tests/pyfunc/test_pyfunc_model_with_type_hints.py

```diff
@@ -0,0 +1,380 @@
+                optional_int=123,
+            ),
+            """
+class Message(pydantic.BaseModel):
+    role: str
+    content: str
+
+
+class CustomExample2(pydantic.BaseModel):
+    custom_field: dict[str, Any]
+    messages: list[Message]
+    optional_int: Optional[int] = None
+""",
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    ("model_type", "has_input_example"),
+    # if python_model is callable, input_example should be provided
+    [("callable", True), ("python_model", True), ("python_model", False)],
+)
+def test_pyfunc_model_with_type_hints_code_based_logging(
+    tmp_path, type_hint_example, model_type, has_input_example
+):
+    kwargs = save_model_file_for_code_based_logging(
+        type_hint_example.type_hint,
+        tmp_path,
+        model_type,
+        type_hint_example.extra_def,
+    )
+    input_example = type_hint_example.input_example
+    if has_input_example:
+        kwargs["input_example"] = input_example
+
+    with mlflow.start_run():
+        model_info = mlflow.pyfunc.log_model("test_model", **kwargs)
+
+    assert model_info.signature is not None
+    assert model_info.signature._is_signature_from_type_hint is True
+    pyfunc_model = mlflow.pyfunc.load_model(model_info.model_uri)
+    assert pyfunc_model.predict(input_example) == input_example
+
+
+def test_functional_python_model_only_input_type_hints():
+    def python_model(x: list[str]):
+        return x
+
+    with mlflow.start_run():
+        model_info = mlflow.pyfunc.log_model(
+            "model", python_model=python_model, input_example=["a"]
+        )
+    assert model_info.signature.inputs == Schema([ColSpec(type=Array(DataType.string))])
+    assert model_info.signature.outputs is None
+
+
+def test_functional_python_model_only_output_type_hints():
+    def python_model(x) -> list[str]:
+        return x
+
+    with mlflow.start_run():
+        model_info = mlflow.pyfunc.log_model(
+            "model", python_model=python_model, input_example=["a"]
+        )
+    assert model_info.signature is None
+
+
+class CallableObject:
+    def __call__(self, x: list[str]) -> list[str]:
+        return x
+
+
+def test_functional_python_model_callable_object():
+    with mlflow.start_run():
+        model_info = mlflow.pyfunc.log_model(
+            "model", python_model=CallableObject(), input_example=["a"]
+        )
+    assert model_info.signature.inputs == Schema([ColSpec(type=Array(DataType.string))])
+    assert model_info.signature.outputs == Schema([ColSpec(type=Array(DataType.string))])
+    loaded_model = mlflow.pyfunc.load_model(model_info.model_uri)
+    assert loaded_model.predict(["a", "b"]) == ["a", "b"]
+
+
+def test_invalid_type_hint_in_python_model():
+    class MyModel(mlflow.pyfunc.PythonModel):
+        def predict(self, model_input: list[object], params=None) -> str:
+            return model_input[0]
+
+    with mlflow.start_run():
+        with mock.patch("mlflow.models.signature.warnings.warn") as mock_warning:
```

Comment:
```suggestion
        with pytest.warns(...):
```

can we use pytest.warns?